Support Sparse CSR transpose. Fix clang-tidy warnings.

pearu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #70582
* #70581
* #70580



cc @nikitaved @pearu @cpuhrsch

Differential Revision: [D33414446](https://our.internmc.facebook.com/intern/diff/D33414446)

pearu committed:
Support Sparse CSR transpose. Fix clang-tidy warnings.

[ghstack-poisoned]

pearu commented:
This PR replaces https://github.com/pytorch/pytorch/pull/68883 and https://github.com/pytorch/pytorch/pull/68938

cpuhrsch commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1650,6 +1685,14 @@ Tensor transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
+  if (self.is_sparse_csr()) {
+    // Sparse CSR transpose is a copy operation as the values of
+    // transposed CSR tensor are permuted values of the input CSR
+    // tensor.
```
NIT: You could add this to the other sparse branch as well or unify them and then dispatch to COO or CSR as needed.

pearu replied:
Done.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33414446).

cpuhrsch commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1562,6 +1562,41 @@ static inline Tensor & sparse_transpose_(Tensor & self, int64_t dim0, int64_t di
 
+static inline Tensor sparse_csr_transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse_csr());
+  auto dim0_ = maybe_wrap_dim(dim0, 2);
```
These are unused which trips up a test

pearu replied:
Fixed. In fact, sparse_csr_transpose does not need the dimensions arguments.

cpuhrsch commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1562,6 +1562,41 @@ static inline Tensor & sparse_transpose_(Tensor & self, int64_t dim0, int64_t di
+static inline Tensor sparse_csr_transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse_csr());
+  auto dim0_ = maybe_wrap_dim(dim0, 2);
+  auto dim1_ = maybe_wrap_dim(dim1, 2);
```
These are unused which trips up a test

pearu replied:
Removed dim0 and dim1 arguments.

pearu committed:
Update on "Support Sparse CSR transpose. Fix clang-tidy warnings."




cc nikitaved pearu cpuhrsch

Differential Revision: [D33414446](https://our.internmc.facebook.com/intern/diff/D33414446)

[ghstack-poisoned]

pearu committed:
Update on "Support Sparse CSR transpose. Fix clang-tidy warnings."




cc nikitaved pearu cpuhrsch

Differential Revision: [D33414446](https://our.internmc.facebook.com/intern/diff/D33414446)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33414446).

cpuhrsch commented:
@pytorchbot ciflow rerun
